sherpa/astro/ui/tests/test_astro_ui_plot.py

```diff
@@ -3579,3 +3579,163 @@ def test_set_plot_opt_explicit_astro():
+def test_plot_bkg_fit_resid_handles_data_log(idval, clean_ui):
+    """Check that log handling is correct: data=log"""
+
+    from matplotlib import pyplot as plt
+
+    setup_example_bkg_model(idval)
+    ui.set_xlog('data')
+    ui.plot_bkg_fit_resid(idval)
+
+    fig = plt.gcf()
+    axes = fig.axes
+    assert len(axes) == 2
+    assert axes[0].xaxis.get_label().get_text() == ''
+
+    assert axes[0].xaxis.get_scale() == 'linear'
+    assert axes[0].yaxis.get_scale() == 'linear'
+
+    assert axes[1].xaxis.get_scale() == 'linear'
+    assert axes[1].yaxis.get_scale() == 'linear'
+
+
+@requires_pylab
+@pytest.mark.parametrize("idval", [None, "bob"])
+def test_plot_bkg_fit_resid_handles_bkg_log(idval, clean_ui):
+    """Check that log handling is correct: bkg=log"""
+
+    from matplotlib import pyplot as plt
+
+    setup_example_bkg_model(idval)
+    ui.set_xlog('bkg')  # NOTE: not 'data'
+    ui.plot_bkg_fit_resid(idval)
+
+    fig = plt.gcf()
+    axes = fig.axes
+    assert len(axes) == 2
+    assert axes[0].xaxis.get_label().get_text() == ''
+
+    assert axes[0].xaxis.get_scale() == 'log'
+    assert axes[0].yaxis.get_scale() == 'linear'
+
+    assert axes[1].xaxis.get_scale() == 'log'
+    assert axes[1].yaxis.get_scale() == 'linear'
+
+
+@requires_pylab
+@pytest.mark.parametrize("idval", [None, "bob"])
+def test_plot_bkg_fit_resid_handles_resid_log(idval, clean_astro_ui):
+    """Check that log handling is correct: resid=log"""
+
+    from matplotlib import pyplot as plt
+
+    setup_example_bkg_model(idval)
+    ui.set_xlog('resid')
+    ui.plot_bkg_fit_resid(idval)
+
+    fig = plt.gcf()
+    axes = fig.axes
+    assert len(axes) == 2
+    assert axes[0].xaxis.get_label().get_text() == ''
+
+    assert axes[0].xaxis.get_scale() == 'linear'
+    assert axes[0].yaxis.get_scale() == 'linear'
+
+    assert axes[1].xaxis.get_scale() == 'linear'
+    assert axes[1].yaxis.get_scale() == 'linear'
+
+
+@requires_pylab
+@pytest.mark.parametrize("idval", [None, "bob"])
+def test_plot_bkg_fit_resid_handles_bkgresid_log(idval, clean_ui):
+    """Check that log handling is correct: bkgresid=log
+
+    We need to decide whether we want the residual setting to override
+    the linear display of the fit plot here. At present the code is
+    that if resid has xlog set then both will be drawn logged (since
+    the X axis is shared via a sharex=True argument to plt.subplots)
+    but we may decide this should change.
+
+    """
+
+    from matplotlib import pyplot as plt
+
+    setup_example_bkg_model(idval)
+    ui.set_xlog('bkgresid')  # NOTE: not 'resid'
+    ui.plot_bkg_fit_resid(idval)
+
+    fig = plt.gcf()
+    axes = fig.axes
+    assert len(axes) == 2
+    assert axes[0].xaxis.get_label().get_text() == ''
+
+    assert axes[0].xaxis.get_scale() == 'log'
+    assert axes[0].yaxis.get_scale() == 'linear'
+
+    assert axes[1].xaxis.get_scale() == 'log'
+    assert axes[1].yaxis.get_scale() == 'linear'
```

Comment:
These are a number of almost identical tests. Maybe they should be parameterized along one more axes?
Something like (untested, not sure about the syntax for passing multiple parnames to parameterize. I know I've done it, I just don't recall right now).
```
@requires_pylab
@pytest.mark.parametrize("idval", [None, "bob"])
@pytest.mark.parametrize("setxlog, xloglin", [('data', 'linear'), ('bkg', 'log'), ('resid', 'linear'), ('bkgresid', 'log'])
def test_plot_bkg_fit_resid_handles_bkgresid_log(idval, clean_ui):
    """Check that log handling is correct:
    MORE GENERAL COMMENT HERE
    """

    from matplotlib import pyplot as plt

    setup_example_bkg_model(idval)
    ui.set_xlog(setxlog)  # NOTE: not 'resid'
    ui.plot_bkg_fit_resid(idval)

    fig = plt.gcf()
    axes = fig.axes
    assert len(axes) == 2
    assert axes[0].xaxis.get_label().get_text() == ''

    assert axes[0].xaxis.get_scale() == xloglin
    assert axes[0].yaxis.get_scale() == 'linear'

    assert axes[1].xaxis.get_scale() == xloglin
    assert axes[1].yaxis.get_scale() == 'linear'
```
That makes the tests easier to read and thus maintain.

Reply:
FYI, I'm pretty sure you need to remove the space to get "setxlog,xloglin" because I think the parser complains with spaces, and you need to add setclog and xloglin parameters to the function (I don't think the relative ordering matters but I try to follow the order of the parametrize call.

It is always easier to copy-n-paste, particluarly when you don't know for sure how the code is meant to behave!

Reply:
I agree that it's easier to write by copy&paste, but it's often easier to maintain when it's shorter. When I read through this I was wondering: How do these tests differ? What's different and what's the same? It's there something subtle that I'm missing? With a parameterization, it's easier to see that these are the same test, just applied to different plots.

But either works and you are working on plotting, so up to you!

Reply:
Doing the refactoring has uncovered a subtle bug - I should have been using clean_astro_ui not clean_ui (well, I had caught this in 71e18221a404a37e3ceba1aadbc6d84a80cd2639 but I thought that I'd pushed it through to the first commit). That fixes one of the caveats I think,

Reply:
Done.